Include/internal/pycore_pymem.h

```diff
@@ -91,6 +91,73 @@ PyAPI_FUNC(int) _PyMem_GetAllocatorName(
 PyAPI_FUNC(int) _PyMem_SetupAllocators(PyMemAllocatorName allocator);
 
 
+#if WITH_FREELISTS
+/* Free lists.
+ *
+ * Free lists have a pointer to their first entry and
+ * the amount of space available allowing fast checks
+ * for emptiness and fullness.
+ * When empty they are half filled and when full they are
+ * completely emptied. This helps the underlying allocator
+ * avoid fragmentation and helps performance.
+ */
+
+typedef struct _freelist {
+    void *ptr;
+    uint32_t space;
+    uint16_t size;
+    uint16_t capacity;
+} _PyFreeList;
+
+extern void *_PyFreeList_HalfFillAndAllocate(_PyFreeList *list);
+extern void _PyFreeList_FreeToFull(_PyFreeList *list, void *ptr);
+
+
+static inline void *
+_PyFreeList_Alloc(_PyFreeList *list) {
+    if (list->ptr != NULL) {
+#ifdef Py_STATS
+        if (_py_stats) _py_stats->object_stats.from_generic_freelist++;
+#endif
+        void *result = list->ptr;
+        list->ptr = *((void **)result);
+        list->space++;
+        return result;
+    }
+#ifdef Py_STATS
+        if (_py_stats) _py_stats->object_stats.generic_freelist_empty++;
+#endif
+    return _PyFreeList_HalfFillAndAllocate(list);
+}
+
+static inline void
+_PyFreeList_Free(_PyFreeList *list, void *ptr) {
+    if (list->space) {
+#ifdef Py_STATS
+        if (_py_stats) _py_stats->object_stats.to_generic_freelist++;
+#endif
+        *((void **)ptr) = list->ptr;
+        list->ptr = ptr;
+        list->space--;
+        return;
+    }
+#ifdef Py_STATS
+        if (_py_stats) _py_stats->object_stats.generic_freelist_full++;
+#endif
+    _PyFreeList_FreeToFull(list, ptr);
```

Comment:
Why don't we just PyObject_Free here and leave the freelist as it is?

Reply:
If a freelist is full it implies one of two things:
* that the freelist is too small, or
* that there is a long sequence of frees.

Assuming the freelist is large enough to the "steady state" of mixed allocations and frees, then we are looking at a long sequence of (mostly) frees.
In which case it makes sense to hand back a bunch of objects to the underlying allocator and avoid the overhead of constantly hitting the slow path.